sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,1061 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_usbdev.h"
+
+#include "sw/device/lib/base/bitfield.h"
+
+#include "usbdev_regs.h"  // Generated.
+
+/**
+ * Definition in the header file (and probably other places) must be updated if
+ * there is a hardware change.
+ */
+_Static_assert(USBDEV_NUM_ENDPOINTS == USBDEV_PARAM_NENDPOINTS,
+               "Mismatch in number of endpoints");
+
+/**
+ * Max packet size is equal to the size of buffer entries.
+ */
+#define USBDEV_BUFFER_ENTRY_SIZE_BYTES USBDEV_MAX_PACKET_SIZE
+
+/**
+ * Bit index for the state of read buffer operation in
+ * dif_usbdev_t.active_buffer_ops`. See the definition of `buffer_op_init`.
+ */
+#define USBDEV_READ_BUFFER_OP_BIT_INDEX 15
+
+/**
+ * Constants used to indicate that a buffer pool is full or empty.
+ */
+#define BUFFER_POOL_FULL (USBDEV_NUM_BUFFERS - 1)
+#define BUFFER_POOL_EMPTY -1
+
+/**
+ * Hardware information for endpoints.
+ */
+typedef struct endpoint_hw_info {
+  uint32_t config_in_reg_offset;
+  uint8_t bit_index;
+} endpoint_hw_info_t;
+
+/**
+ * Helper macro to define an `endpoint_hw_info_t` entry for endpoint N.
+ *
+ * Note: This uses the bit indices of `USBDEV_IN_SENT` register for the sake
+ * of conciseness because other endpoint registers use the same layout.
+ */
+#define ENDPOINT_HW_INFO_ENTRY(N)                                 \
+  [N] = {.config_in_reg_offset = USBDEV_CONFIGIN##N##_REG_OFFSET, \
+         .bit_index = USBDEV_IN_SENT_SENT##N}
```

Comment:
This is unpleasant, but I agree that it's a better algernative than risking getting the constants wrong.